lib/shoryuken/launcher.rb

```diff
@@ -28,12 +28,25 @@ def stop
 
       initiate_stop
 
+      stop_new_dispatching
+      await_dispatching_in_progress
+
       executor.shutdown
       executor.wait_for_termination
     end
 
     private
 
+    def stop_new_dispatching
+      @managers.each do |manager|
+        manager.stop_new_dispatching
+      end
+    end
+
+    def await_dispatching_in_progress
+      sleep(Shoryuken.await_dispatching_timeout) if @managers.any? { |manager| manager.dispatching_in_progress? }
```

Comment:
@nelhefni Lemme know if you're interested in making this change yourself. Otherwise, I'm happy to merge as-is and leave the optimization up to another contributor.

Reply:
@cjlarose, could you please check the last commit? @tomgi implemented the change using mutex

Reply:
Awesome. I'll take a look!

Reply:
Hey @nelhefni, sorry it's taken a while to get back to you with feedback. What I've been struggling with is that I'm not 100% convinced the new mutex-based mechanism is correct.

It seems possible that the thread executing `Manager#dispatch_loop` can find that `@stop_new_dispatching.true?` is not true (we haven't yet stopped dispatching), just before `Manager#stop_new_dispatching` and `Manager#await_dispatching_in_progress` are invoked. But then that same thread that was executing `#dispatch_loop` would still post the `dispatch` job to the executor pool (assuming the executor is not yet shut down). That would allow the manager to fetch messages from the queue without giving the opportunity to actually process the message on the executor pool.

```ruby
    def dispatch_loop
      return unless running?
      return if @stop_new_dispatching.true?
      # imagine that TSTP is received here
      @executor.post { dispatch }
    end
```

This might be better explained with an explicit listing of the code invoked by each thread:

| Main thread (signal handler)    | Executor pool worker thread 1 | Executor pool worker thread 2         |
|---------------------------------|-------------------------------|---------------------------------------|
|                                 | `@stop _new_dispatching.true?`  |                                       |
|                                 | => `false`                     |                                       |
| `@stop_new_dispatching.make_true` |                               |                                       |
| `@dispatch_mutex.synchronize {}`  |                               |                                       |
| => (executes immediately)       |                               |                                       |
|                                 | `@executor.post { dispatch }`   |                                       |
|                                 |                               | Acquires `@dispatch_mutex`              |
|                                 |                               | `@fetcher.fetch(queue, ready)`          |
| `executor.shutdown`               |                               |                                       |
|                                 |                               | `Manager#assign`                        |
|                                 |                               | => fails because executor is shutdown |

What do you think?

Reply:
hi @cjlarose, good catch! @tomgi and I implemented a fix, moving `@dispatch_mutex` to the `dispatch_loop` method, and introducing a Queue as a way to signal when the mutex should be released from another thread running  `dispatch_loop`

![image](https://www.planttext.com/api/plantuml/svg/bLAnRi8m4Dtz5IUcTY0IoORKD4l73XKhKdHABiIgyHdx0lJlQq88IPV2OfrExzrxxyv54FJIxOrIBXQwqWwjmE8JjGLfFM6zjIz1xomQQD7MXlph0Z10a3dZuqnL9-p1CPkjFR7_9Z-CW5LFsSlg2KuwSD8dEDd0oMQSpTUcTtm8IaKFopm-wre7XrArfM5szpPMXqvx2b2CW7qdT5POYJwYK9zYX1T1s45fwLJUyDhk4l4TlSDR3WsQG3TuEehU1y09jK3Jvt5R-qUrz6G80vNN2qHQkinZZdL3bocnkiwDXXtRg7H5JALk-OQswq8BmlygJHmsEZRCpqC6QvhjwQxxo5xJ4PzDDvLD8rkEK3x6arzG47ya10QsIP8eLP2jur_-1G00)

Reply:
Ok awesome. I think the new mechanism correctly prevents shutting down the executor except _between_ dispatch runs, which is exactly what we want. Thanks for the nice sequence diagram, too! That helps me understand what's going on and will serve as documentation moving forward.

I'll play around with this a little bit locally, but I think this is mostly good-to-go.